Sources/SourceKitLSP/Swift/Refactoring.swift

```diff
@@ -0,0 +1,130 @@
+  ///
+  /// - Parameters:
+  ///   - title: The title of the refactoring action.
+  ///   - dict: Response dictionary to extract information from.
+  ///   - snapshot: The snapshot that triggered the `semantic_refactoring` request.
+  ///   - keys: The sourcekitd key set to use for looking up into `dict`.
+  init?(_ title: String, _ dict: SKDResponseDictionary, _ snapshot: DocumentSnapshot, _ keys: sourcekitd_api_keys) {
+    guard let categorizedEdits: SKDResponseArray = dict[keys.categorizedEdits] else {
+      logger.error("categorizedEdits doesn't exist in response dictionary")
+      return nil
+    }
+
+    var refactoringEdits = [RefactoringEdit]()
+
+    categorizedEdits.forEach { _, categorizedEdit in
+      guard let edits: SKDResponseArray = categorizedEdit[keys.edits] else {
+        logger.error("edits doesn't exist in categorizedEdit dictionary")
+        return true
+      }
+      edits.forEach { _, edit in
+        // The LSP is zero based, but semantic_refactoring is one based.
+        guard let startLine: Int = edit[keys.line],
+          let startColumn: Int = edit[keys.column],
+          let endLine: Int = edit[keys.endLine],
+          let endColumn: Int = edit[keys.endColumn],
+          let text: String = edit[keys.text]
+        else {
+          logger.error("Failed to deserialise edit dictionary containing values: \(edit)")
+          return true  // continue
+        }
+
+        let startPosition = snapshot.positionOf(
+          zeroBasedLine: startLine - 1,
+          utf8Column: startColumn - 1
+        )
+        let endPosition = snapshot.positionOf(
+          zeroBasedLine: endLine - 1,
+          utf8Column: endColumn - 1
+        )
+        // Snippets are only supported in code completion.
+        // Remove SourceKit placeholders in refactoring actions because they
+        // can't be represented in the editor properly.
+        let textWithSnippets = rewriteSourceKitPlaceholders(in: text, clientSupportsSnippets: false)
+        refactoringEdits.append(
+          RefactoringEdit(
+            range: startPosition..<endPosition,
+            newText: textWithSnippets,
+            bufferName: edit[keys.bufferName] ?? ""
+          )
+        )
+        return true
+      }
+      return true
+    }
+
+    guard !refactoringEdits.isEmpty else {
+      logger.error("No refactors found")
+      return nil
+    }
+
+    self.init(title: title, uri: snapshot.uri, refactoringEdits: refactoringEdits)
+  }
+}
+
+extension SwiftLanguageService {
+  /// Provides detailed information about the result of a specific refactoring
+  /// operation.
+  ///
+  /// Wraps the information returned by sourcekitd's `semantic_refactoring`
+  /// request, such as the necessary edits and placeholder locations.
+  ///
+  /// - Parameters:
+  ///   - refactorCommand: The semantic `RefactorCommand` that triggered this request.
+  /// - Returns:
+  ///   - The `RefactoringResponse` as `T`
```

Comment:
Also, I’m not entirely sure what you mean by _The `RefactoringResponse` as `T`_. I would just go for something like: _The response of the refactoring_.

---

Nit: You don’t need a sub-bullet point for the description of the return value.

```suggestion
  /// - Returns: The `RefactoringResponse` as `T`
```